algoliasearch/src/main/java/com/algolia/search/saas/BaseAPIClient.java

```diff
@@ -469,102 +474,75 @@ private synchronized JSONObject _request(Method m, String url, String json, List
-            } catch (IOException e) {
-                // on error continue on the next host
-                addError(errors, host, e);
-                continue;
-            }
+                // read response
+                int code = hostConnection.getResponseCode();
+                final boolean codeIsError = code / 100 != 2;
+                InputStream stream = codeIsError ?
+                        hostConnection.getErrorStream() : hostConnection.getInputStream();
 
-            InputStream stream = hostConnection.getErrorStream(); // Response is in ErrorStream unless code = 200
-            if (code / 100 == 2) {
-                // OK
-                try {
-                    stream = hostConnection.getInputStream();
-                } catch (IOException e) {
-                    addError(errors, host, e);
-                    continue;
-                }
-            } else if (code / 100 == 4) {
-                String message = "Error detected in backend";
-                try {
-                    message = _getAnswerJSONObject(stream).getString("message");
-                } catch (IOException e) {
-                    addError(errors, host, e);
-                    continue;
-                } catch (JSONException e) {
-                    throw new AlgoliaException("JSON decode error:" + e.getMessage());
-                }
-                consumeQuietly(hostConnection);
-                throw new AlgoliaException(message);
-            } else {
-                try {
-                    errors.put(host, _toCharArray(stream));
-                } catch (IOException e) {
-                    errors.put(host, String.valueOf(code));
-                }
-                consumeQuietly(hostConnection);
-                // KO, continue
-                continue;
-            }
-            try {
+                final byte[] rawResponse;
                 String encoding = hostConnection.getContentEncoding();
                 if (encoding != null && encoding.contains("gzip")) {
-                    return _toByteArray(new GZIPInputStream(stream));
+                    rawResponse = _toByteArray(new GZIPInputStream(stream));
                 } else {
-                    return _toByteArray(stream);
+                    rawResponse = _toByteArray(stream);
+                }
+
+                // handle http errors
+                if (codeIsError) {
+                    if (code / 100 == 4) {
+                        String message = _getJSONObject(rawResponse).getString("message");
+                        consumeQuietly(hostConnection);
+                        throw new AlgoliaException(message);
+                    } else {
+                        final String errorMessage = _toCharArray(stream);
+                        errors.put(host, errorMessage);
+                        consumeQuietly(hostConnection);
+                        throw new IOException(errorMessage);
```

Comment:
I don't like the idea of throwing an `IOException` here, because:

1. It is **not** an I/O error, but a server application error!
2. It is using an exception for flow control, which is kind of an anti-pattern.

Why not just adding the error and continuing the loop?